### PR TITLE
Circular json can possibly hang

### LINKED ISSUE #778: Circular json can possibly hang.
<a href="https://github.com/rhalff"></img></a> **Issue by [rhalff](https://github.com/rhalff)**
_Wednesday Nov 02, 2016 at 02:55 GMT_
_Originally opened as https://github.com/storybooks/storybook-addon-actions/issues/26_

----

It can be replicated by replacing `circular` with something like `window` within:

 https://github.com/kadirahq/storybook-addon-actions/blob/master/.storybook/stories.js#L20

The click on the circular button will cause the action to never be logged.

I've tried replacing `json-stringify-safe` with `circular-json` which seems to behave better.
But even then after the 4th click `circular-json` will throw a circular reference error.

I now know how to circumvent this by just not logging something with circular references, however when this bug is encountered it's not immediately clear the hang originates from within storybook which makes one try to debug the application code itself first.

I think replacing json-stringify-safe with circular-json would already be an improvement.

### PR DESCRIPTION
**Pull request by rhalff**
* replaced `json-stringify-safe` with a custom version of [JSON-js](https://github.com/douglascrockford/JSON-js/blob/master/cycle.js)'s decycle and retrocycle.
* hide non-enumerable properties:
  It makes no sense to render these while the logged objects come directly from JSON.parse.
  The non-enumerability is used to set some properties to restore the original constructor names. `react-inspector` will then use these to render the correct labels.
* custom node renderer to restore the constructor names
* if an object is marked as cyclic (by decycle) no attempt will be made to test for deep equality.
* added a max depth of 15, when this level is reached traversal stops and three extra dots are added to the label: ` [<name>...]`
* sort object keys

Perhaps the documentation should make clear that whatever is logged is **not** the original object.
This is due to the objects being stringified first and then restored again.

fixes #778

### CODE REVIEW

File: addons/actions/src/preview.test.js
Comment (ndelangen) [inline addons/actions/src/preview.test.js]: Ha, that's kind of funny actually!

**Review by ndelangen — APPROVED** [review]
LGTM

File: addons/actions/src/util.js
Comment (ndelangen) [inline addons/actions/src/util.js]: Might this be it's own npm module maybe?

**Comment by ndelangen** [comment]
Thank you for this PR @rhalff ! 🙇 

I have a question about if we could extract the util you added into a separate npm module, maybe now, or later. But I won't block merging for that.

**Comment by ndelangen** [comment]
Waiting for a second review

**Comment by danielduan** [comment]
this might fix - #1839

**Review by danielduan — COMMENTED** [review]
LGTM +1

this feels like a patch version to me. thoughts @ndelangen @shilman?

**Review by danielduan — APPROVED** [review]

**Pull request merged**
